federation/vector_clock.go

```diff
@@ -0,0 +1,35 @@
+package federation
+
+import (
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+type vectorClock map[peer.ID]uint64
+
+func newVectorClock() vectorClock {
+	return make(map[peer.ID]uint64)
+}
+
+func (vc vectorClock) clock(id peer.ID) (uint64, bool) {
+	clock, found := vc[id]
+	return clock, found
+}
+
+func (vc vectorClock) tick(id peer.ID) uint64 {
+	vc[id] = vc[id] + 1
+	return vc[id]
+
+}
+
+func (vc vectorClock) untick(id peer.ID) uint64 {
```

Comment:
can you add comments on what these methods do.

i'm uncertain why we want to run the clock backwards - that's generally a much harder state machine to model when allowed

Reply:
Will add comments. 

The reason for this is to handle error cases correctly, where taking a snapshot might fail before it is stored in the snapshot chain, and if that happens we don't want to leave the clock incremented. 

Please let me know if you have any alternative suggestions on how to atomically record a snapshot without untick.

Reply:
the other pattern might be to make a "copy" at the higher tick, but not commit it back to the snapshot until reaching consensus/finality. - it's nice to have the assumption that the 'consensus' / 'committed' vector clock will monotonically increase